src/fec_base.h

```diff
@@ -787,53 +788,53 @@ void FecCode<T>::decode_prepare(
  */
 template <typename T>
 void FecCode<T>::decode_apply(
-    DecodeContext<T>* context,
+    const DecodeContext<T>& context,
     vec::Vector<T>* output,
     vec::Vector<T>* words)
 {
-    unsigned len_2k = context->get_len_2k();
+    unsigned len_2k = context.get_len_2k();
 
-    vec::Vector<T>* fragments_ids = context->fragments_ids;
-    std::shared_ptr<vec::Poly<T>> inv_A_i = context->inv_A_i;
-    std::shared_ptr<vec::Poly<T>> A_fft_2k = context->A_fft_2k;
-    std::shared_ptr<vec::Poly<T>> poly1_n = context->poly1_n;
-    std::shared_ptr<vec::Poly<T>> poly2_n = context->poly2_n;
-    std::shared_ptr<vec::Poly<T>> poly1_2k = context->poly1_2k;
-    std::shared_ptr<vec::Poly<T>> poly2_2k = context->poly2_2k;
+    vec::Vector<T>* fragments_ids = context.get_fragments_id();
```

Comment:
I think these should be reference, because you shouldn't have a raw pointer from a `std::unique_ptr`